knowledge-asset-tools.js

```diff
@@ -130,7 +136,166 @@ async function peerId2Hash(peerId) {
 }
 
+function groupNquadsBySubject(nquadsArray) {
+  const store = new N3.Store();
+  const parser = new N3.Parser({ format: "star" });
+
+  nquadsArray.forEach((quad) => {
+    try {
+      const parsedQuad = parser.parse(quad);
+      parsedQuad.forEach((quad) => store.addQuad(quad));
+    } catch (error) {
+      console.error("Error parsing quad:", quad);
+      console.error(error);
+    }
+  });
+
+  function groupTriples(triples) {
+    const grouped = {};
+
+    triples.forEach((quad) => {
+      let subject = quad.subject ? quad.subject : null;
+      const predicate = quad.predicate.value;
+      let object = quad.object;
+
+      if (subject && subject.termType === "Quad") {
+        const nestedSubjectKey = `<<${subject.subject.value}> <${subject.predicate.value}> ${subject.object.value}>>`;
+
+        if (!grouped[nestedSubjectKey]) {
+          grouped[nestedSubjectKey] = [];
+        }
+
+        grouped[nestedSubjectKey].push([predicate, object.value]);
+      } else {
+        object = object ? object.value : "";
+
+        if (!grouped[subject.value]) {
+          grouped[subject.value] = [];
+        }
+
+        grouped[subject.value].push([predicate, object]);
+      }
+    });
+
+    return grouped;
+  }
+
+  const allTriples = store.getQuads(null, null, null, null);
+  return groupTriples(allTriples);
+}
+
+function countDistinctSubjects(nquadsArray) {
+  const store = new N3.Store();
+  const parser = new N3.Parser({ format: "star" });
+
+  nquadsArray.forEach((quad) => {
+    try {
+      const parsedQuad = parser.parse(quad);
+      parsedQuad.forEach((quad) => store.addQuad(quad));
+    } catch (error) {
+      console.error("Error parsing quad:", quad);
+      console.error(error);
+    }
+  });
+
+  const subjects = new Set();
 
+  store.getQuads(null, null, null, null).forEach((quad) => {
+    const subject = quad.subject.value;
+    subjects.add(subject);
+  });
+
+  return subjects.size;
+}
+
+function filterTriplesByAnnotation(
+  nquadsArray,
+  annotationPredicate = null,
+  annotationValue = null
+) {
+  const store = new N3.Store();
+  const parser = new N3.Parser({ format: "star" });
+
+  nquadsArray.forEach((quad) => {
+    try {
+      const parsedQuad = parser.parse(quad);
+      parsedQuad.forEach((quad) => store.addQuad(quad));
+    } catch (error) {
+      console.error("Error parsing quad:", quad);
+      console.error(error);
+    }
+  });
+
+  const filteredSubjects = [];
+
+  store.getQuads(null, null, null, null).forEach((quad) => {
+    const subject = quad.subject;
+    const predicate = quad.predicate.value;
+    const object = quad.object ? quad.object.value : null;
+
+    if (subject.termType === "Quad") {
+      const nestedSubject = subject.subject.value;
+      const nestedPredicate = subject.predicate.value;
+      const nestedObject = subject.object.value;
+
+      let matches = false;
+      if (annotationPredicate && annotationValue) {
```

Comment:
think at the beginning we should check if both are defined, and error if they are not imo